Add way of identifying and loading a TLA module's dependencies from the file system

shonfeder opened the pull request:
Closes #31

Two principles changes are introduced here:

- The `Source` class is reworked as a representation in the client of the `SourceOption` structures we use for inputs in Apalache. This allows a type-safe way of supporting inputs from different data sources (strings and files on disk, currently, but could be expanded to taking python dicts) and different formats (JSON and TLA+ currently).
- Methods are added to produce a `Source` from a single file on disk, that will determine possible dependencies and search for these in the same directory. This is to satisfy #31.

The rest of the changes are only updating tests or making small adaptations.

Commit:
Rework `Source` data structure

`Source` was previously an abstraction letting us load the content of
files into strings. This change makes it instead a representation of the
input sources that Apalache needs in order to parse, typecheck, or model
check specs.

The bulk of the complexity here comes from functions used to find the names
of module dependencies from within the text of a TLA file.

Commit:
Use the reworked source type throughout

shonfeder commented:
- [x] I need to add support for transitive deps.

Commit:
Fix formatting

Includes removal of `isort`, which was getting fights with black.

Commit:
Support lookup of transitive deps

thpani commented on chai/source.py:
```diff
@@ -0,0 +1,169 @@
+        return Source(
+            source=p.read_text(),
+            aux=[s.read_text() for s in aux],
+            format=p.suffix.lstrip("."),
```
I'd add a comment here, for readability:

```suggestion
            format=p.suffix.lstrip("."),  # last filename suffix, without leading "."
```

thpani commented on chai/source.py:
```diff
@@ -0,0 +1,169 @@
+    def __init__(
+        self,
+        source: str,
+        *,
```
I'm confused about the asterisk here; it's not bound to a name... is it meant to unpack arguments?
I haven't found an invocation that would use this.

shonfeder replied:
This forces all other arguments to be passed via keyword: see https://peps.python.org/pep-3102/

thpani replied:
I've never seen this before. Could you add a comment?

```suggestion
        *,  # force all other arguments to be keyword-only
```

shonfeder replied:
Yep good idea. Thanks.

thpani commented on chai/source.py:
```diff
@@ -0,0 +1,169 @@
+    A source from which the client can load data
+    """
+
+    # Supported inputs to derive a `Source`
+    # Strings should be a
+    Input = Union[str, Path]
```
I don't understand the "Strings should be a" part 😅

shonfeder replied:
I think this type alias should have been removed actually, as it's not used now.

thpani commented on chai/source.py:
```diff
@@ -0,0 +1,169 @@
+            return source.read_text()
+        else:
+            raise ValueError(
+                "Source can only be construced from a str or a Path,"
```
I think mentioning `Source` here is confusing, since this returns only a `str`?

thpani commented on chai/source.py:
```diff
@@ -0,0 +1,169 @@
+    ]
+
+
+class Source:
+    """
+    A source from which the client can load data
+    """
```
It took me some time to understand the interface of this class.
afaict this supports two different kinds of callers?

1. The caller knows what kind of `Input` it's loading. These use `Source/__init__` and `of_file[_load_deps]`.
2. The caller doesn't and has to go through `load_input`, then call `Source/__init__` on the results.

It seems that we could have a more uniform interface by providing a `@classmethod` that dynamically dispatches based on the `Input` type like `load_input`, but actually returns a `Source`?

Alternatively, I'd suggest to expand the docstring:

```suggestion
class Source:
    """
    In-memory representation of a TLA+ module and it's dependencies as strings.

    To initialize from filesystem objects, use the class methods `of_file`
    and `of_file_load_deps`.
    If the kind of `Input` is unknown, use `load_input` to transparently
    convert an `Input` into a string.
    """
```

thpani replied:
Yet another thought: Should `Input` be a class that has a `load() -> str` method?

shonfeder replied:
Good questions here.
 
> It seems that we could have a more uniform interface by providing a @classmethod that dynamically dispatches based on the Input type like load_input, but actually returns a Source?

Two things have motivated moving away from an overloaded constructor working on both `str`s and `Path`s:

- Andrey voiced a preference against that approach in our sync with MBT a few weeks back 
- `str` inputs *must* be accompanied by a format specification (because IMO it would be too much magic to try to infer the format by inspecting the string), but this is unnecessary and even introduces the potential for errors to support formats for files, for which the format is derived from the extension.

So we really need two separate constructors to safely construct the source.

The `load_input` really doesn't have anything to do with the `Source` class now, it's just a convenience function, and maybe should be made private... I'll think it over a bit. 

I'll incorporate something like your docstring update. Thanks!

shonfeder replied:
Sorry for the messy work here: it turns out `load_input` was really only "needed" due to the earlier design decisions made for the `trans_explorer`, prior to work started on the `cmd_executor`. The subsequent work has changed the RPC approach quite a bit, so we should followup with reworking the `trans_explorer` code. I've opened an issue to track this:  #33

shonfeder replied:
See 

- [c88eae8](https://github.com/informalsystems/apalache-chai/pull/32/commits/c88eae8e7eea8c4d219489f63d764042986464e1) for the "deprecation" of the `load_input` method 
- 85fa4ccf5807e5155acfaa4280abda4f5bbd26a7 for inclusion of your suggested docstring

thpani replied:
Looks good!
I think I was really just confused by the 3 class methods and how to combine them to construct a source; thanks for addressing this!

shonfeder replied:
You'd definitely caught a messy bit needing cleaning up. Thanks!

thpani commented on .github/workflows/main.yml:
```diff
@@ -24,10 +24,9 @@ jobs:
+      - run: python -m pip install black flake8
       - run: flake8 .
       - run: black . --check
-      - run: isort .
```
No more `isort`? 😢

shonfeder replied:
I was hitting conflicts with Black, so black would reformat the imports one way that isort would not like, then isort would format them another that black wouldn't like.

I removed isort in anger, tho I should have searched for a way of reconciling them :D

thpani replied:
Aaah, I hit that as well! You can add

```
[tool.isort]
profile = "black"
```

to `pyproject.toml`.

shonfeder replied:
oh great. thanks!

thpani reviewed (approved):
The dependency resolution looks really powerful!
Have you tested it on any "real" sources, like from `test/tla` or `tla-files`?

We should consider also using it in the cloud client!

I have a question about the `Source` class interface tho 😬

Commit:
Update chai/source.py

Co-authored-by: Thomas Pani <thomas.pani@gmail.com>

Commit:
Update chai/source.py

Co-authored-by: Thomas Pani <thomas.pani@gmail.com>

Commit:
Move load_input into trans_explorer

This is atavistic, and is only really needed for support for the WIP
state of the `trans_explorer`. This should be replaced by use of a
`Source` class instead, but work on that will left for a followup, as it
also calls for changes on the Scala RPC side.

Tracked in  #33

Commit:
Add @thpani's docstrings

Commit:
Add isort back in

shonfeder commented:
Thanks again for the helpful review! :pray:
